### PR TITLE
Additional async-chores for SMTP/IMAP health checks

### PR DESCRIPTION
**Pull request by peterwurzinger**
**What this PR does / why we need it**:
Some chores regarding async methods available in .NET Standard 2.0 and > .NET 5 and passing the cancellation token in operations that could potentially block as discussed in https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/1739#issuecomment-1529389296

**Which issue(s) this PR fixes**:
None, but related to #1738

**Special notes for your reviewers**:
This PR is based on branch https://github.com/peterwurzinger/AspNetCore.Diagnostics.HealthChecks/tree/fix/smtp-timeout
It doesn't make sense to base it on main, since it would result in merge conflicts, assuming that the fix in #1738 would get merged first anyways.

**Does this PR introduce a user-facing change?**:
No, it only introduces binary breaking changes which shouldn't be a problem according to https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/pull/1739#issuecomment-1475151121

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Unit tests passing
- [x] End-to-end tests passing

### CODE REVIEW

**Comment by sungam3r** [comment]
OK, resolve conflicts and I will review.

**Comment by peterwurzinger** [comment]
@sungam3r Resolved

File: src/HealthChecks.Network/Core/ImapConnection.cs [hunk 1]
```diff
@@ -74,20 +74,16 @@ private async Task<bool> UpgradeToSecureConnectionAsync()
         }
     }
 
-#pragma warning disable IDE1006 // Naming Styles
-    public async Task<bool> SelectFolder(string folder) //TODO: public API change
-#pragma warning restore IDE1006 // Naming Styles
+    public async Task<bool> SelectFolderAsync(string folder, CancellationToken cancellationToken = default) //TODO: public API change
```
Comment (sungam3r) [inline src/HealthChecks.Network/Core/ImapConnection.cs:77]:
```suggestion
    public async Task<bool> SelectFolderAsync(string folder, CancellationToken cancellationToken = default)
```

File: src/HealthChecks.Network/Core/ImapConnection.cs [hunk 2]
```diff
@@ -74,20 +74,16 @@ private async Task<bool> UpgradeToSecureConnectionAsync()
...
+        var result = await ExecuteCommandAsync(ImapCommands.SelectFolder(folder), cancellationToken).ConfigureAwait(false);
 
         //Double check, some servers sometimes include a last line with a & OK appending extra info when command fails
         return result.Contains(ImapResponse.OK) && !result.Contains(ImapResponse.ERROR);
     }
 
-#pragma warning disable IDE1006 // Naming Styles
-    public async Task<string> GetFolders() //TODO: public API change
-#pragma warning restore IDE1006 // Naming Styles
+    public async Task<string> GetFoldersAsync(CancellationToken cancellationToken = default) //TODO: public API change
```
Comment (sungam3r) [inline src/HealthChecks.Network/Core/ImapConnection.cs:85]:
```suggestion
    public async Task<string> GetFoldersAsync(CancellationToken cancellationToken = default)
```

File: src/HealthChecks.Network/Core/MailConnection.cs [hunk 3]
```diff
@@ -77,23 +86,35 @@ protected SslStream GetSSLStream(Stream stream)
         }
     }
 
-#pragma warning disable IDE1006 // Naming Styles
-    protected async Task<string> ExecuteCommand(string command) //TODO: rename public API
-#pragma warning restore IDE1006 // Naming Styles
+    protected async Task<string> ExecuteCommandAsync(string command, CancellationToken cancellationToken) //TODO: rename public API
```
Comment (sungam3r) [inline src/HealthChecks.Network/Core/MailConnection.cs:89]:
```suggestion
    protected async Task<string> ExecuteCommandAsync(string command, CancellationToken cancellationToken)
```

**Comment by sungam3r** [comment]
Thanks.

**Comment by sungam3r** [comment]
This PR was merged without running tests, see #1018.

**Comment by peterwurzinger** [comment]
> This PR was merged without running tests, see #1018.

Is there something broken or to do? I ran manual tests and unit tests against a SMTP/IMAP implementation (https://github.com/rnwood/smtp4dev), which doesn't support (STAR)TLS, but basic SMTP and IMAP functionalityand coudln't observe any errors that would not be expected.

**Comment by sungam3r** [comment]
Test step in network CI worflow is disabled https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/master/.github/workflows/healthchecks_network_ci.yml#L83-L97

I think this is because not properly configured services https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/master/.github/workflows/healthchecks_network_ci.yml#L32-L62

**Pull request merged**